markdown.go

```diff
@@ -0,0 +1,321 @@
+	fmt.Fprintln(w, "| Flag | Default | Usage |")
+	fmt.Fprintln(w, "| --- | --- | --- |")
+
+	for _, fs := range byName {
+		// Collect all flag aliases (usually a short one and a plain one, like -v / --verbose)
+		formattedFlags := ""
+		for i, f := range fs {
+			if i > 0 {
+				formattedFlags += ", "
+			}
+			if len(f.Name) == 1 {
+				formattedFlags += fmt.Sprintf("`-%s`", f.Name)
+			} else {
+				formattedFlags += fmt.Sprintf("`--%s`", f.Name)
+			}
+		}
+		// display all the flags aliases and the default value and description of the shortest one.
+		// Escape Markdown in description in order to display it cleanly in the final documentation.
+		fmt.Fprintf(w, "| %s | %s | %s |\n", formattedFlags,
+			EscapeMarkdown(fs[0].DefValue),
+			strings.ReplaceAll(EscapeMarkdown(fs[0].Usage), "\n", " "),
+		)
+	}
+	fmt.Fprintln(w)
+}
+
+// flagsByLength is a slice of flags implementing sort.Interface,
+// sorting primarily by the length of the flag, and secondarily
+// alphabetically.
+type flagsByLength []*gnuflag.Flag
+
+func (f flagsByLength) Less(i, j int) bool {
+	s1, s2 := f[i].Name, f[j].Name
+	if len(s1) != len(s2) {
+		return len(s1) < len(s2)
+	}
+	return s1 < s2
+}
+func (f flagsByLength) Swap(i, j int) {
+	f[i], f[j] = f[j], f[i]
+}
+func (f flagsByLength) Len() int {
+	return len(f)
+}
+
+// flagsByName is a slice of slices of flags implementing sort.Interface,
+// alphabetically sorting by the name of the first flag in each slice.
+type flagsByName [][]*gnuflag.Flag
+
+func (f flagsByName) Less(i, j int) bool {
+	return f[i][0].Name < f[j][0].Name
+}
+func (f flagsByName) Swap(i, j int) {
+	f[i], f[j] = f[j], f[i]
+}
+func (f flagsByName) Len() int {
+	return len(f)
+}
+
+func printSubcommands(
+	w io.Writer,
+	subcommands map[string]string,
+	linkForSubcommand func(string) string,
+) {
+	sorted := []string{}
+	for name := range subcommands {
+		if isDefaultCommand(name) {
+			continue
+		}
+		sorted = append(sorted, name)
+	}
+	sort.Strings(sorted)
+
+	if len(sorted) > 0 {
+		fmt.Fprintln(w, "## Subcommands")
+		for _, name := range sorted {
+			fmt.Fprint(w, "- ")
+			fmt.Fprint(w, markdownLink(name, linkForSubcommand))
+			fmt.Fprintln(w)
+		}
+		fmt.Fprintln(w)
+	}
+}
+
+// markdownLink uses the provided linker function to generate a Markdown
+// hyperlink for the given key. It attempts to call the linker function on the
+// given key to get the link target. If the function is nil or the output is
+// empty, just the key (as a non-link) will be returned.
+func markdownLink(key string, linker func(string) string) string {
+	var target string
+	if linker != nil {
+		target = linker(key)
+	}
+
+	if target == "" {
+		// We don't have a link target for this key, so just return the key.
+		return key
+	} else {
+		return fmt.Sprintf("[%s](%s)", key, target)
+	}
+}
+
+// EscapeMarkdown returns a copy of the input string, in which any special
+// Markdown characters (e.g. < > |) are escaped.
+func EscapeMarkdown(raw string) string {
```

Comment:
Looks like EscapeMarkdown needs a test.
Also it seems like it doesn't escape all `markdown` https://spec.commonmark.org/current/ so perhaps the name should be more specific to the task at hand.

Also perhaps look at use test fixtures in a testdata directory to compare output to (rather than having to escape again in Go strings)

Reply:
EscapeMarkdown does a partial escape of certain Markdown characters that can cause problems in outputted Markdown docs. I wanna be conservative here and only escape characters which we know to be an issue. There are many special Markdown characters that we specifically **don't** want to escape, for example hyphens `-`, we **want** hyphenated plain-text lists to display as formatted bulletpoints in Markdown.

Maybe the name's not 100% accurate but I can't think of anything better right now. Perhaps `PartialEscapeMarkdown` but that doesn't really feel like an improvement. The doc comment could be improved at least, we don't escape "*any* special Markdown characters", only a subset of them.

FYI this is not new code, it was moved here from `documentation.go` as I think it makes more sense here. This is tested in the documentation suite already.